lib/logstash/codecs/kms.rb

```diff
@@ -0,0 +1,140 @@
+  # See http://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html
+  config :secret_key, :validate => :string, :default => nil
+
+  # An optional AWS profile to use.
+  # See http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html#cli-multiple-profiles
+  config :aws_profile, :validate => :string, :default => nil
+
+  # An encryption context to use to encrypting and decrypting data.
+  # When ecrypting, the encryption context is sent as a part of the payload
+  # When decrypting, if any of the key-value pairs specified in encryption_context is
+  # missing from the payload's encryption context, then decryption will fail.
+  # See: http://docs.aws.amazon.com/kms/latest/developerguide/encryption-context.html
+  config :encryption_context, :validate => :hash, :default => {}
+
+  # If set, incoming messages that cannot be decrypted will try to pass the original message
+  # through this plugin's codec. By default, undecryptable messages will fail immediately.
+
+  # This is useful to avoid issues while you are migrating to an encrypted transport.
+  # For example, if you are currently sending unencrypted data, but want to start sending KMS encrypted data,
+  # there will be a window during your deployment where old hosts are sending decrypted data while new hosts
+  # are sending encrypted data.
+  # You can handle this case by setting fallback_if_invalid_format to true.
+  #
+  # The following example will pass messages that fail to be decrypted through the json codec.
+  # [source,ruby]
+  #     stdin {
+  #         codec => kms {
+  #             codec => "json"
+  #             fallback_if_invalid_format => true
+  #         }
+  #     }
+  #
+  config :fallback_if_invalid_format, :validate => :boolean, :default => false
+
+  def register
+    @crypto_client = com.amazonaws.encryptionsdk::AwsCrypto.new
+
+    credentials = com.amazonaws.auth::DefaultAWSCredentialsProviderChain.new
+    if @access_key and @access_key
+      credentials = com.amazonaws.auth::AWSStaticCredentialsProvider.new(
+        com.amazonaws.auth::BasicAWSCredentials.new(
+          @access_key,
+          @secret_key
+        )
+      )
+    elsif @aws_profile
+      credentials = com.amazonaws.auth.profile::ProfileCredentialsProvider.new(@aws_profile)
+    end
+
+    @key_provider = com.amazonaws.encryptionsdk.kms::KmsMasterKeyProvider.new(
+      credentials,
+      com.amazonaws.regions::RegionUtils::getRegion(@region),
+      com.amazonaws::ClientConfiguration.new,
+      @key_ids
+    )
+  end # def register
+
+  def decode(data)
+    begin
+      response = get_crypto_client().decryptData(@key_provider, data.to_java_bytes)
+      context = response.getEncryptionContext()
+      @encryption_context.each do |key, value|
+        if not context.containsKey(key) or context[key] != value
+          raise RuntimeError.new('Encryption context does not match expected. Recieved context: ' + context.to_s)
+        end
+      end
```

Comment:
yeah. The SDK adds additional metadata into the encryption context as well. The point is to just ensure that fields you expect to be there are matching, additional context that you don't validate is OK, and actually should be allowed otherwise things wont work.